lib/pure/terminal.nim

```diff
@@ -406,22 +406,74 @@ proc isatty*(f: File): bool =
 
   result = isatty(getFileHandle(f)) != 0'i32
 
-proc styledEchoProcessArg(s: string) = write stdout, s
-proc styledEchoProcessArg(style: Style) = setStyle({style})
-proc styledEchoProcessArg(style: set[Style]) = setStyle style
-proc styledEchoProcessArg(color: ForegroundColor) = setForegroundColor color
-proc styledEchoProcessArg(color: BackgroundColor) = setBackgroundColor color
-
-macro styledEcho*(m: varargs[expr]): stmt =
-  ## to be documented.
-  let m = callsite()
-  result = newNimNode(nnkStmtList)
-
-  for i in countup(1, m.len - 1):
-    result.add(newCall(bindSym"styledEchoProcessArg", m[i]))
-
-  result.add(newCall(bindSym"write", bindSym"stdout", newStrLitNode("\n")))
-  result.add(newCall(bindSym"resetAttributes"))
+type
+  StyledMsgKind* = enum
+    msgText
+    msgStyle
+    msgForegroundColor
+    msgBackgroundColor
+    msgReset
+  StyledMsg* {.final, acyclic.} = object
+    case kind*: StyledMsgKind
+    of msgText:
+      text*: string
+    of msgStyle:
+      style*: Style
+    of msgForegroundColor:
+      fg*: ForegroundColor
+    of msgBackgroundColor:
+      bg*: BackgroundColor
+    of msgReset:
+      discard
+
+const
+  ResetStyle* = StyledMsg(kind: msgReset)
+
+converter toStyledMsg*(text: string): StyledMsg =
+  StyledMsg(kind: msgText, text: text)
+converter toStyledMsg*(style: Style): StyledMsg =
+  StyledMsg(kind: msgStyle, style: style)
+converter toStyledMsg*(fg: ForegroundColor): StyledMsg =
+  StyledMsg(kind: msgForegroundColor, fg: fg)
+converter toStyledMsg*(bg: BackgroundColor): StyledMsg =
+  StyledMsg(kind: msgBackgroundColor, bg: bg)
+converter toStyledMsg*(msg: StyledMsg): StyledMsg = msg
+
+proc styledWriteln*(output: File, msgs: varargs[StyledMsg, toStyledMsg]) =
```

Comment:
it's cool but the old macro was way more efficient. ;-) But I can see why it doesn't scale to what you need in msgs.nim. Will sleep over it.